scene/resources/environment.cpp

```diff
@@ -891,23 +894,58 @@ float Environment::get_adjustment_saturation() const {
 	return adjustment_saturation;
 }
 
-void Environment::set_adjustment_color_correction(const Ref<Texture2D> &p_ramp) {
+void Environment::set_adjustment_color_correction(Ref<Texture> p_ramp) {
 	adjustment_color_correction = p_ramp;
 	_update_adjustment();
 }
 
-Ref<Texture2D> Environment::get_adjustment_color_correction() const {
+Ref<Texture> Environment::get_adjustment_color_correction() const {
 	return adjustment_color_correction;
 }
 
 void Environment::_update_adjustment() {
+	Ref<GradientTexture> grad_tex = adjustment_color_correction;
+	adjustment_color_correction_cache = adjustment_color_correction;
+	Ref<Texture2D> texture_2d = adjustment_color_correction_cache;
+	Ref<ImageTexture3D> texture_3d_image = adjustment_color_correction_cache;
+	Ref<StreamTexture3D> texture_3d_stream = adjustment_color_correction_cache;
+	RID color_correction = adjustment_color_correction_cache.is_valid() ? adjustment_color_correction_cache->get_rid() : RID();
+	if (texture_2d.is_valid()) {
```

Comment:
You are reloading the texture every time ``update_adjustment()`` is called. This means if you use any of the sliders, you may be loading the 3D texture thousands of times.

Reply:
Is delaying the load by 0.3 of a second a good alternative?

Reply:
This technique was done for screenshots to avoid capturing menus.

Reply:
Delaying isn't a good idea, again because this function is called when sliders are used for BCS.

Reply:
Any suggestions? I'm out of ideas.

Reply:
@clayjohn Mentioned using different shader versions. Still unresolved, but we can retest with the new options.

![image](https://user-images.githubusercontent.com/32321/99198594-94a97d00-274e-11eb-9c9d-f9bb77036da6.png)